odl/solvers/mlem/mlem.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Maximum Likelihood Expectation Maximization algorithm."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+__all__ = ('mlem_poisson',)
+
+
+def mlem_poisson(op, x, data, iter=1, noise='poisson', callback=None):
+
+    """Implementation of Maximum Likelihood Expectation Maximation algorithm.
+
+    This method computes the MLEM estimates for Poisson noise.
+
+    Parameters
+    ----------
+    op : `Operator`
+        Operator in the inverse problem. It has to have adjoint that is called
+        via `op.adjoint`
+
+    x : `element` of the domain of ``op``
+        Vector to which the result is written. Its initial value is
+        used as starting point of the iteration, and its values are
+        updated in each iteration step.
+        The initial value of `x` should be non negative.
+
+    data : `element` of the range of ``op``
+        Right-hand side of the equation defining the inverse problem
+
+    niter : `int`, optional
+        Maximum number of iterations
+
+    callback : `callable`, optional
+        Object executing code per iteration, e.g. plotting each iterate
+
+    noise : `str`, optional
+        Implementation back-end for the noise.
+    """
+    if x <= 0:
+        raise ValueError('Initial value of x needs to be positive')
+
+    noise = str(noise).lower()
+
+    if noise.startswith('poisson'):
```

Comment:
This can be otimized quite a bit:

```python
eps = 1e-8
norm = np.maximum(op.adjoint(op.range.one()), eps)
tmp_dom = op.domain.element()
tmp_ran = op.range.element()

for _ in range(iter):
    op(x, out=tmp_ran)
    np.maximum(tmp_ran, eps, out=tmp_ran)

    op.adjoint(tmp_ran, out=tmp_dom)
    tmp_dom /= norm

    x *= tmp_dom
```